cups/ipp-file.c

```diff
@@ -647,31 +647,37 @@ parse_value(_ipp_file_t      *f,	/* I  - IPP data file */
     	{
     		if(value[0]=='<')				/* Input is binary(in form of hex) values*/
     		{
-    			memmove(value, value+1, strlen(value));		/* Eliminate the '<' sign */
-    			char value_concat[50000];		/* Concatenated string with hexadecimal values without whitespace */
-    			int starting_line_number = f->linenum;
-    			while(value[strlen(value)-1] != '>') 
-    			{
-    				strcat(value_concat,value);
-    				if (!_ippFileReadToken(f, value, sizeof(value)))
-					  {
-					    report_error(f, v, user_data, "hexadecimal value not terminated starting line %d of \"%s\".", starting_line_number, f->filename);
-					    return (0);
-					  }
-    			}
-    			value[(strlen(value)-1)]='\0'; 		/* Eliminate the last '>' sign */
-    			strcat(value_concat,value);			/* Final concatenation for hexadecimal value to be complete*/
-    			int i;                         /* Iterating variable*/
-          for(i=0;i<strlen(value_concat);i++)		/* Sanity Check*/
-    			{
-    				if(!(value_concat[i] >= '0' && value_concat[i]<= '9') || (value_concat[i]>='a' && value_concat[i]<='f'))
-    				{
-    					report_error(f, v, user_data, "Bad hexadecimal value \"%s\" on line %d of \"%s\".", value_concat, starting_line_number, f->filename);
-	    				return (0);
-    				}
-    			}
-    			return (ippSetOctetString(ipp, attr, element, value_concat, (int)strlen(value_concat))); 
-    		}
+          unsigned char data[32767], *dataptr = data; /*data: decoded hex, dataptr:iterating pointer */
+          char *valptr = value + 1; /*Value iterating pointer*/
+          while (1)
+            {
+              while (isxdigit(valptr[0]) && isxdigit(valptr[1]))              
+              {
+                 char c = valptr[0], d=valptr[1];
```

Comment:
Need "tolower" here, as "isxdigit" allows upper and lowercase letters.

Reply:
Fixed!